### PR TITLE
add custom samplers

### PR DESCRIPTION
**Pull request by parissashahabi**
It needs some adjustments, but it would be helpful if you could review it and let me know if I should separate some of the classes into different files or move anything into other folders.

@philippkolbe

### CODE REVIEW

File: src/neural_bandits/utils/data_sampler.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    def _get_iterator(self) -> Iterator[int]:
+        """Returns an iterator that yields indices in random order."""
+        indices = torch.randperm(
+            len(self.data_source), generator=self.generator, dtype=torch.int64
+        ).tolist()
+
+        return iter(indices)
+
+
+class SortedDataSampler(AbstractDataSampler):
```
Comment (philippkolbe) [inline src/neural_bandits/utils/data_sampler.py:63]: can you add some tests for the SortedDataSampler? 😄

File: src/neural_bandits/utils/data_sampler.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def __iter__(self) -> Iterator[int]:
+        return self._get_iterator()
+
+    @abstractmethod
+    def _get_iterator(self) -> Iterator[int]:
+        """Core sampling logic to be implemented by subclasses."""
+        raise NotImplementedError
+
+
+class RandomDataSampler(AbstractDataSampler):
```
[the remark this answers is not in the document]
Comment (philippkolbe) [reply src/neural_bandits/utils/data_sampler.py:40]: The thing is that in the DataLoader you can already specify if you want to shuffle or not so I don't think we actually need the RandomDataSampler.

File: examples/neuralucb.ipynb
Comment (philippkolbe) [inline examples/neuralucb.ipynb:1]: You are going to get really annoying merge conflicts with my interface PR on this one. If you do, I think you can just remove these changes, especially if you have unit tests.

File: src/neural_bandits/utils/data_sampler.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Iterator, Optional
+
+import torch
+from torch.utils.data import Sampler
+
+from neural_bandits.datasets.abstract_dataset import AbstractDataset
+
+
+# TODO: Correct docstring format
```
Comment (philippkolbe) [inline src/neural_bandits/utils/data_sampler.py:10]: docstring format looks good. you can remove the TODO 👍

File: src/neural_bandits/utils/data_sampler.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    def __len__(self) -> int:
+        return len(self.data_source)
+
+    def __iter__(self) -> Iterator[int]:
+        return self._get_iterator()
+
+    @abstractmethod
+    def _get_iterator(self) -> Iterator[int]:
+        """Core sampling logic to be implemented by subclasses."""
+        raise NotImplementedError
```
Comment (philippkolbe) [inline src/neural_bandits/utils/data_sampler.py:37]: nit: The abstractmethod already makes sure that if this is not implemented by subclasses that it will throw an error.
Usually, I just write "pass" in the body of abstract methods.

File: src/neural_bandits/utils/data_sampler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (philippkolbe) [reply src/neural_bandits/utils/data_sampler.py:40]: Then I'm also questioning if we need the AbstractDataSampler if we only have the SortedDataSampler😢 
What do you think @parissashahabi?

[the remark this answers is not in the document]
Comment (parissashahabi) [reply src/neural_bandits/utils/data_sampler.py:40]: I wasn’t sure if we’d have other types of samplers in the future, so I added the abstract sampler and the random data sampler to keep things consistent 🤷🏻‍♀️

File: src/neural_bandits/utils/data_sampler.py (hunk 1 above)
Comment (parissashahabi) [reply src/neural_bandits/utils/data_sampler.py:63]: I've added the [issue](https://github.com/neural-bandits/calvera/issues/111)

File: src/neural_bandits/utils/data_sampler.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    Args:
+        data_source: Dataset to sample from
+    """
+
+    def __init__(
+        self,
+        data_source: AbstractDataset,
+    ) -> None:
+        self.data_source = data_source
```
[the remark this answers is not in the document]
Comment (parissashahabi) [reply src/neural_bandits/utils/data_sampler.py:26]: I think I should not 🤔:

``` python
class Sampler(Generic[_T_co]):
    r"""Base class for all Samplers.

    Every Sampler subclass has to provide an :meth:`__iter__` method, providing a
    way to iterate over indices or lists of indices (batches) of dataset elements,
    and may provide a :meth:`__len__` method that returns the length of the returned iterators.

    Args:
        data_source (Dataset): This argument is not used and will be removed in 2.2.0.
            You may still have custom implementation that utilizes it.

    Example:
        >>> # xdoctest: +SKIP
        >>> class AccedingSequenceLengthSampler(Sampler[int]):
        >>>     def __init__(self, data: List[str]) -> None:
        >>>         self.data = data
        >>>
        >>>     def __len__(self) -> int:
        >>>         return len(self.data)
        >>>
        >>>     def __iter__(self) -> Iterator[int]:
        >>>         sizes = torch.tensor([len(x) for x in self.data])
        >>>         yield from torch.argsort(sizes).tolist()
        >>>
        >>> class AccedingSequenceLengthBatchSampler(Sampler[List[int]]):
        >>>     def __init__(self, data: List[str], batch_size: int) -> None:
        >>>         self.data = data
        >>>         self.batch_size = batch_size
        >>>
        >>>     def __len__(self) -> int:
        >>>         return (len(self.data) + self.batch_size - 1) // self.batch_size
        >>>
        >>>     def __iter__(self) -> Iterator[List[int]]:
        >>>         sizes = torch.tensor([len(x) for x in self.data])
        >>>         for batch in torch.chunk(torch.argsort(sizes), len(self)):
        >>>             yield batch.tolist()

    .. note:: The :meth:`__len__` method isn't strictly required by
              :class:`~torch.utils.data.DataLoader`, but is expected in any
              calculation involving the length of a :class:`~torch.utils.data.DataLoader`.
    """

    def __init__(self, data_source: Optional[Sized] = None) -> None:
        if data_source is not None:
            import warnings

            warnings.warn(
                "`data_source` argument is not used and will be removed in 2.2.0."
                "You may still have custom implementation that utilizes it."
            )
            ```

**Review by philippkolbe — APPROVED** [review]
looks great
